kobo/hub/models.py

```diff
@@ -260,7 +304,49 @@ def _get_relative_log_path(self, name):
             raise RuntimeError("Invalid log normpath.")
         return log_path
 
+    def get_chunk(self, name, offset=0, length=-1):
+        """Returns a sequence of bytes from the named log.
+
+        offset -- optional start index of read
+        length -- optional length; if omitted, read to end of file
+
+        Will try to avoid returning a chunk ending in the middle of
+        a single UTF-8 character, therefore the returned length may
+        be less than requested."""
+
+        name = self._get_relative_log_path(name)
+
+        if name in self.cache:
+            # already loaded so no point in reading again
+            end = None if length < 0 else offset + length
+            return self.cache[name][offset:end]
+
+        # not loaded; read just this part
+        log_path = self._get_absolute_log_path(name)
+        log_file = None
+        try:
+            if os.path.isfile(log_path):
+                log_file = open(log_path, 'rb')
```

Comment:
A problem I have with this is that no specific encoding is enforced on the log files.

From reviewing the code writing to logs, the status seems to be:

- whatever raw bytes are written to stdout/stderr during a task execution will be appended to stdout.log
- when an exception occurs in a task, no encoding is explicitly applied to the strings written to error.log and traceback.log
- the upload_task_log XML-RPC method, which could be used by applications built on kobo to upload custom logs, appends any binary data to a log file

There's nothing in the model which can tell me what's the expected encoding of a log file, so I may be unable to use codecs.open safely, and I don't know if it'll be safe to introduce a new requirement that all log files in kobo must be UTF-8, or must use the system's default locale (there could be existing kobo systems out there with logs violating this).

_utf8_chunk on the other hand ought to be safe - if you use UTF-8 (recommended) then it makes things work better by aligning chunks at character boundaries.  If you don't use UTF-8 then it's encoding-neutral, you get exactly the byte range you asked for.

What do you think?